src/main/java/software/amazon/cloudformation/proxy/package-info.java

```diff
@@ -0,0 +1,461 @@
+/**
+ * This package provide facilities to make it easy to work against AWS APIs that
+ * are eventually consistent for applying resource state. Developers need to
+ * sequence APIs calls in order to effectively apply state. When dependent
+ * resources are not available, e.g. when associating KMS Key with CloudWatch
+ * Log group, the key might not have propogated to region leading to failures to
+ * associate with the log group. This framework provides developers the
+ * facilities to deal with all of these situations using a simple functional
+ * model.
+ * <p>
+ * The framework handles the following for developers:
+ * <ol>
+ * <li>Taking care of communicating all progress events consistently with
+ * CloudFormation including appropriate handshakes</li>
+ * <li>Handles all client side error reporting correctly with appropriate
+ * failure codes and errors messages</li>
+ * <li>Handles common AWS service exceptions like Throttle and several other
+ * retry errors based on HTTP codes</li>
+ * <li>Can auto-renew STS token expiry transparently if needed or schedule
+ * callback from CloudFormation requesting new credentials</li>
+ * <li>Automatically handles AWS Lambda runtime constraints and schedules
+ * callbacks when we near max computation time. Auto adjusts behavior based on
+ * the runtime the handler is bound other modes for operating</li>
+ * <li>Memoize all service call requests, responses, stabilization and retry
+ * handling that ensures continuations of handler compute start from where we
+ * left off for callbacks.</li>
+ * <li>Provides default flat retry model for calls of 5 seconds with max timeout
+ * of 2 minutes for all API calls. Allows for sensible defaults to be specified
+ * for retry strategy by integrator on a per API call basis. Allows framework to
+ * dynamically override backoff behavior for helping with emergencies with
+ * downstream service latency and other situations defined by customer need</li>
+ * <li>Provides developers with a simple sequential pattern of progress chaining
+ * while doing all the heavy lifting above. Provides most common sensible
+ * defaults with override capabilities for integrator to influence as needed.
+ * Covers power users and simple users alike.</li>
+ * <li>Brings consistent model across all service integrations across AWS
+ * services on GitHub hosted code base.</li>
+ * </ol>
+ * <p>
+ * <b><u>Anatomy of an AWS Web Service call</u></b>
+ * <p>
+ * Most AWS web service API calls follows a typical pattern shown below
+ * <ol>
+ * <li>Initiate the call context in preparation of the API</li>
```

Comment:
I think this initiate part is what throws me when I look at the pattern. What actually happens here, it seems like copy-pasta in all implementations.

Reply:
First each initiate call encodes the IAM action for that API calls. E.g.

```java
        return initiator.initiate("logs:DeleteLogGroup")
            .translate(m -> DeleteLogGroupRequest.builder().logGroupName(m.getLogGroupName()).build())
            .call((r, c) -> c.injectCredentialsAndInvokeV2(r, c.client()::deleteLogGroup))
            .done(ignored -> ProgressEvent.success(null, context));

```
The IAM permissions for the call is "logs:DeleteLogGroup". This makes it easier to quickly inspect and find out the IAM permissions in the code base.

Second Initiate allows the developers to easily track request/responses inside the CallbackContext and access them if they need in others parts of the overall chain if needed. E.g.

```java
        return initiator.initiate("networkmanager:CreateGlobalNetwork")
            //
            // Make the request object to create from the model
            //
            .translate(model ->
                CreateGlobalNetworkRequest.builder()
                    .description(model.getDescription())
                    .tags(Utils.cfnTagsToSdkTags(model.getTags()))
                    .build())
            //
            // Make the call the create the global network. Delegate to framework to retry all errors and
            // report all errors as appropriate including service quote exceeded and others.
            //
            .call((r, c) -> {
                CreateGlobalNetworkResponse res = c.injectCredentialsAndInvokeV2(r, c.client()::createGlobalNetwork);
                GlobalNetwork network = res.globalNetwork();
                initiator.getResourceModel().setArn(network.globalNetworkArn());
                initiator.getResourceModel().setId(network.globalNetworkId());
                return res;
            })
            //
            // Check to see if Global Network is available to use directly from the response or
            // stabilize as needed. Update model with Arn and Id as primary identifier on the model
            // object to communicate to CloudFormation about progress
            //
            .stabilize((_request, _response, _client, _model, _context) -> {
                GlobalNetworkState state = _response.globalNetwork().state();
                return state == GlobalNetworkState.AVAILABLE ||
                    Utils.globalNetwork(_client, _response.globalNetwork().globalNetworkId()).state() == GlobalNetworkState.AVAILABLE;
            }).progress();
```

and somewhere else in the chain you needed to access the response object you can 

```java
        CallbackContext cxt = ...
        CreateGlobalNetworkResponse response =  
             (CreateGlobalNetworkResponse) cxt.callGraphs().get("networkmanager:CreateGlobalNetwork.response");
``` 

Can add helper methods on callback context. This helps re-use information from inside the CallContext.